lib/spack/llnl/util/lang.py

```diff
@@ -227,48 +233,203 @@ def list_modules(directory, **kwargs):
+
+    Raises:
+        TypeError: If the class does not have a ``_cmp_iter`` method
 
-       Raises:
-           TypeError: If the class does not have a ``_cmp_key`` method
     """
-    def setter(name, value):
-        value.__name__ = name
-        setattr(cls, name, value)
-
-    if not has_method(cls, '_cmp_key'):
-        raise TypeError("'%s' doesn't define _cmp_key()." % cls.__name__)
-
-    setter('__eq__',
-           lambda s, o:
-           (s is o) or (o is not None and s._cmp_key() == o._cmp_key()))
-    setter('__lt__',
-           lambda s, o: o is not None and s._cmp_key() < o._cmp_key())
-    setter('__le__',
-           lambda s, o: o is not None and s._cmp_key() <= o._cmp_key())
-
-    setter('__ne__',
-           lambda s, o:
-           (s is not o) and (o is None or s._cmp_key() != o._cmp_key()))
-    setter('__gt__',
-           lambda s, o: o is None or s._cmp_key() > o._cmp_key())
-    setter('__ge__',
-           lambda s, o: o is None or s._cmp_key() >= o._cmp_key())
-
-    setter('__hash__', lambda self: hash(self._cmp_key()))
+    if not has_method(cls, "_cmp_iter"):
+        raise TypeError("'%s' doesn't define _cmp_iter()." % cls.__name__)
+
+    # eq and lt are the main comparison implementations
+    def eq(lseq, rseq):
+        liter = lseq()  # call generators
+        riter = rseq()
+
+        # zip_longest is implemented in native code, so use it for speed.
+        # use zip_longest instead of zip because it allows us to tell
+        # which iterator was longer.
+        for left, right in zip_longest(liter, riter, fillvalue=done):
+            if (left is done) or (right is done):
+                return False
+
+            # recursively enumerate any generators, otherwise compare
+            equal = eq(left, right) if callable(left) else left == right
+            if not equal:
+                return False
+
+        return True
+
+    # these are helpers for lt that evaluate which iterator finished first
+    left_shorter  = lambda left, right: left is done
+    right_shorter = lambda left, right: right is done
+
+    # lt is implemented much like eq.
+    def lt(lseq, rseq, op=operator.lt, diff=left_shorter):
+        liter = lseq()
+        riter = rseq()
+
+        for left, right in zip_longest(liter, riter, fillvalue=done):
+            if (left is done) or (right is done):
+                return diff(left, right)
+
+            equal = eq(left, right) if callable(left) else left == right
+            if equal:
+                continue
+            return (
+                lt(left, right, op, diff) if callable(left)
+                else op(left, right)
+            )
+
+        return False  # if equal, return False
+
+    # other comparison operators are implemented in terms of eq and lt
+    def ne(lseq, rseq):
+        return not eq(lseq, rseq)
+
+    def gt(lseq, rseq):
+        return lt(lseq, rseq, operator.gt, right_shorter)
```

Comment:
```suggestion
        return lt(rseq, lseq)
```
Then we can remove op and diff as arguments to `lt`.

Reply:
fixed